spinalcordtoolbox/image.py

```diff
@@ -973,7 +973,17 @@ def get_dimension(im_file, verbose=1):
     ndims = [1, 1, 1, 1]
     pdims = [1, 1, 1, 1]
     data_shape = im_file.header.get_data_shape()
-    zooms = im_file.header.get_zooms()
+    # NB: Nibabel stores dim info as float32. Arithmetic using float32 values can cause issues. So, store the values
+    #     as rounded 64-bit float values, according to the spacing of the value.
+    zooms = []
+    for zoom_f32 in im_file.header.get_zooms():
+        # extract the exponent from the spacing used by the float32 values (i.e. 1e-9 -> '-9')
+        spacing = np.spacing(zoom_f32)
+        spacing_exp = np.floor(np.log10(np.abs(spacing))).astype(int)
+        # convert to python float, then round to remove any spurious digits
+        zoom_f64 = float(zoom_f32)
+        zoom_f64_rounded = round(zoom_f64, abs(spacing_exp) - 1)
```

Comment:
The `abs(spacing_exp)` should actually be `-spacing_exp`. You can check this with an example that has a positive `e` exponent:
- `zoom_f32 = np.float32(1e+30)`
- then `spacing_exp == 22`
- the incorrect result is `round(float(zoom_f32), 21) == 1.0000000150474662e+30`
- the correct result is `round(float(zoom_f32), -23) == 1e+30`
- and this isn't off-by-one, because `round(float(zoom_f32)), -22) == 1.00000002e+30`.

(If you're wondering what rounding to a negative number of digits means: `round(12345, -2) == 12300`)
```suggestion
        zoom_f64_rounded = round(zoom_f64, -spacing_exp - 1)
```

Reply:
Ahhh, good catch! 

I think the reason I got mixed around here is that, because zooms are typically only ever in fractions of mm, I've only ever seen negative exponents? (It would be a bit strange to have the resolution be in meters/kilometers/etc, hehe.)

But, I appreciate the thoroughness and correctness -- you never know what sort of inputs we could get!